junit-platform-surefire-provider/src/main/java/org/junit/platform/surefire/provider/JUnitPlatformProvider.java

```diff
@@ -103,10 +125,53 @@ private void invokeSingleClass(Class<?> testClass, RunListener runListener) {
 		SimpleReportEntry classEntry = new SimpleReportEntry(getClass().getName(), testClass.getName());
 		runListener.testSetStarting(classEntry);
 
-		LauncherDiscoveryRequest discoveryRequest = request().selectors(selectClass(testClass)).build();
+		LauncherDiscoveryRequest discoveryRequest = request().selectors(selectClass(testClass)).filters(
+			includeAndExcludeFilters).build();
 		launcher.execute(discoveryRequest);
 
 		runListener.testSetCompleted(classEntry);
 	}
 
+	private Filter<?>[] getIncludeAndExcludeFilters() {
+		List<Filter<?>> filters = new ArrayList<>();
+
+		Optional<List<String>> includes = getGroupsOrTags(getPropertiesList(INCLUDE_GROUPS),
+			getPropertiesList(INCLUDE_TAGS));
+		if (includes.isPresent()) {
+			filters.add(TagFilter.includeTags(includes.get()));
+		}
+
+		Optional<List<String>> excludes = getGroupsOrTags(getPropertiesList(EXCLUDE_GROUPS),
+			getPropertiesList(EXCLUDE_TAGS));
+		if (excludes.isPresent()) {
+			filters.add(TagFilter.excludeTags(excludes.get()));
+		}
+
+		return filters.toArray(new Filter<?>[filters.size()]);
+	}
+
+	private Optional<List<String>> getPropertiesList(String key) {
+		List<String> compoundProperties = null;
+		String property = parameters.getProviderProperties().get(key);
+		if (property != null) {
+			compoundProperties = Arrays.asList(property.split("[, ]+"));
+		}
+		return Optional.ofNullable(compoundProperties);
+	}
+
+	private Optional<List<String>> getGroupsOrTags(Optional<List<String>> groups, Optional<List<String>> tags) {
+		Optional<List<String>> elements = Optional.empty();
+
+		Preconditions.condition(groups.isPresent() && tags.isPresent(), EXCEPTION_MESSAGE_BOTH_NOT_ALLOWED);
```

Comment:
Yes, `Preconditions.condition()` throw an exception if the first argument is `false`. `!groups.isPresent() || !tags.isPresent()` is only `false` when both `groups` and `tags` are present.